### PR TITLE
Enable pool to return memory to the upstream upon OOM.

### PR DESCRIPTION
**Pull request by mzient**
Signed-off-by: Michał Zientkiewicz <mzient@gmail.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It adds new feature: **returning memory to upstream** needed because we allocate huge memory chunks for tensor lists.

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     * Added an option to get a specific address range from a free list and free tree
     * Upon OOM, run over upstream blocks and return the completely free blocks to the upstream, removing them from free list
     * If the upstream uses paging, it can rearrange physical storage and assign a new continuous virtual address space to the released fragments
 - Affected modules and functionalities:
     * Pooling resource
     * Free lists
     * Test tracking resource
 - Key points relevant for the review:
     *[ Describe here what is the most important part that reviewers should focus on. ]*
 - Validation and testing:
     * Unit tests for free_list
     * Unit test for pooling resource that checks that the deallocation actually happens
 - Documentation (including examples):
     * Doxygen docstrings, when necessary

**JIRA TASK**: DALI-2027 (needed as a fix)

### CODE REVIEW

File: include/dali/core/mm/detail/free_list.h [hunk 1]
```diff
@@ -278,6 +276,43 @@ class best_fit_free_list {
     unused_blocks_ = b;
   }
 
+  /**
+   * @brief Gets a specific chunk of memory from a specific block
+   *
+   * The start and end parameters must lie inside the block.
+   */
+  void *get_from_block(block **pblock, char *start, char *end) {
```
Comment (mzient) [inline include/dali/core/mm/detail/free_list.h:284]: This function is factored out of the old `get` - it's not new code.

File: dali/core/mm/free_list_test.cc [hunk 2]
```diff
@@ -148,6 +170,14 @@ TEST(MMFreeTree, PutGet) {
   TestCoalescingPutGet<free_tree>();
 }
 
+TEST(MMCoalescingFreeList, RemoveIf) {
+  TestCoalescingRemoveIf<coalescing_free_list>();
```
Comment (JanuszL) [inline dali/core/mm/free_list_test.cc:174]: Shouldn't you test `best_fit_free_list` as well?

Comment (mzient) [reply dali/core/mm/free_list_test.cc:174]: Well, that would require a different test. The implementation is almost trivial, though, and I don't plan to use it - it's there more or less for symmetry.

File: include/dali/core/mm/detail/free_list.h [hunk 3]
```diff
@@ -478,6 +539,62 @@ class free_tree {
...
+   */
+  void *get_specific_block(char *start, char *end) {
+    assert(end > start);
+    if (by_addr_.empty())
+      return nullptr;
+    size_t size = end - start;
+
+    auto it = by_addr_.lower_bound(start);
+    if (it == by_addr_.end() || it->first > start)
+      it--;
```
[the remark this answers is not in the document]
Comment (mzient) [reply include/dali/core/mm/detail/free_list.h:555]: Good point.

[the remark this answers is not in the document]
Comment (mzient) [reply include/dali/core/mm/detail/free_list.h:555]: Fixed.

File: dali/core/mm/free_list_test.cc (hunk 2 above)
Comment (mzient) [reply dali/core/mm/free_list_test.cc:208]: Done.

**Review by JanuszL — APPROVED** [review]

**Review by banasraf — APPROVED** [review]

**Pull request merged**